core/state_transition.go

```diff
@@ -318,7 +318,8 @@ func (st *StateTransition) TransitionDb() (ret []byte, usedGas uint64, failed bo
 
 	err = st.payFees()
 	if err != nil {
-		log.Error("Transaction failed to buy gas", "err", err, "gas", gas)
+		// Expected error when called by EstimateGas() for accounts with balance < binary search hi
+		log.Warn("Transaction failed to buy gas. This is expected when invoked by EstimateGas() for accounts with small balances", "err", err, "gas", gas)
```

Comment:
So this is an error because if a transaction is being processed here, it is assumed that it was already validated by the transaction pool, in which case this should never happen. We are breaking that assumption in the `EstimateGas` function. I think the best way to handle this would be to run the transaction through transaction pool validation, either in its entirety or just to validate that the transaction can pay its fees, before applying the state transition.
Here is the `validateTx` function on `TxPool`, which you obviously can't call directly as it stands https://github.com/celo-org/celo-blockchain/blob/f86d6733274bb43377584cb9475a2fe46e17f368/core/tx_pool.go#L607
And here is the part that validates balances to ensure fees can be paid https://github.com/celo-org/celo-blockchain/blob/f86d6733274bb43377584cb9475a2fe46e17f368/core/tx_pool.go#L1201
Two courses of action would either be to export the former, or just call the latter if that turns out to be inappropriate. The call should probably occur here-ish https://github.com/celo-org/celo-blockchain/blob/f86d6733274bb43377584cb9475a2fe46e17f368/internal/ethapi/api.go#L740

If this seems onerous, please leave this log statement as is and copy the above into an issue to address later. And feel free to assign me to it.

Reply:
Set back to original logging and created the issue!